Test metadataSchemaValidation fails if table gpkg_metadata contains multiple values with at least one not NMIS valid entry

Failing test:  org.opengis.cite.gpkg12.nsg.metadata.MetadataTests#metadataSchemaValidation()

Requirement is not described in the test. Probably it is:
```
NSG Req 12: A GeoPackage SHALL contain one complete NMIS metadata XML instance 
document that describes the entire GeoPackage in a gpkg_metadata table row with the 
contents shown in below.
```
Test  executes the following SQL-Statement `SELECT metadata from gpkg_metadata;`. All results are validated against the nmis.xsd.

If the SQL Statement returns e.g. the following three rows the test fails:
```
d1eea157-4f94-4852-9fde-bf8d9327e6d1
------------------------------------------------------------
<?xml version="1.0" encoding="UTF-8"?>
<nas:MD_Metadata xmlns:gco="http://www.isotc211.org/2005/gco" xmlns:gmd="http://www.isotc211.org/2005/gmd" xmlns:gmi="http://www.isotc211.org/2005/gmi" xmlns:gml="http://www.opengis.net/gml/3.2" xmlns:ism="urn:us:gov:ic:ism" xmlns:nas="http://metadata.ces.mil/dse/ns/GSIP/5.0/nas" xmlns:ntk="urn:us:gov:ic:ntk" xmlns:swe="http://www.opengis.net/swe/2.0" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" ism:DESVersion="9" ism:classification="U" ism:createDate="2018-04-17" ism:ownerProducer="USA" ism:resourceElement="true" ntk:DESVersion="7" xsi:schemaLocation="http://metadata.ces.mil/dse/ns/GSIP/5.0/nas ./nas/nmis.xsd">
<gmd:hierarchyLevel>
...
</nas:MD_Metadata>
------------------------------------------------------------
ALL
```

Currently "java.lang.AssertionError: Failed to read XML resource. Content is not allowed in prolog." is reported if one metadata entry is a simple string (not XML).

--- a/pom.xml
+++ b/pom.xml
@@ -73,6 +73,12 @@
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <version>2.21.0</version>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <repositories>

--- a/src/main/java/org/opengis/cite/gpkg12/nsg/metadata/MetadataTests.java
+++ b/src/main/java/org/opengis/cite/gpkg12/nsg/metadata/MetadataTests.java
@@ -1,14 +1,12 @@
 package org.opengis.cite.gpkg12.nsg.metadata;
 
-import static org.opengis.cite.gpkg12.ETSAssert.assertSchemaValid;
+import static org.testng.Assert.assertTrue;
 
 import java.io.ByteArrayInputStream;
 import java.net.URL;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
-import java.util.logging.Level;
-import java.util.logging.Logger;
 
 import javax.xml.XMLConstants;
 import javax.xml.transform.Source;
@@ -18,12 +16,23 @@ import javax.xml.validation.SchemaFactory;
 import javax.xml.validation.Validator;
 
 import org.opengis.cite.gpkg12.CommonFixture;
+import org.opengis.cite.validation.ValidationErrorHandler;
 import org.testng.SkipException;
+import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
+import org.xml.sax.SAXException;
 
 public class MetadataTests extends CommonFixture {
 
-    private final static Logger LOG = Logger.getLogger( MetadataTests.class.getName() );
+    private Schema schema;
+
+    @BeforeClass
+    public void initSchema()
+                            throws SAXException {
+        SchemaFactory schemaFactory = SchemaFactory.newInstance( XMLConstants.W3C_XML_SCHEMA_NS_URI );
+        URL resource = getClass().getClassLoader().getResource( "org/opengis/cite/gpkg12/nsg/metadata/NMIS_v2.X_Schema/nas/nmis.xsd" );
+        this.schema = schemaFactory.newSchema( resource );
+    }
 
     /**
      * Validate metadata against NMIS xsd https://nsgreg.nga.mil/doc/view?i=2491
@@ -34,30 +43,45 @@ public class MetadataTests extends CommonFixture {
     @Test(description = "Validate against NMIS schema")
     public void metadataSchemaValidation()
                             throws SQLException {
-        Schema schema = createSchema();
-        if ( schema == null )
+        if ( this.schema == null )
             throw new SkipException( "Schema required for validation could not be loaded." );
 
         try (final Statement statement = this.databaseConnection.createStatement();
                                 final ResultSet resultSet = statement.executeQuery( "SELECT metadata FROM gpkg_metadata;" )) {
-            while ( resultSet.next() ) {
-                String xmlResult = resultSet.getString( "metadata" );
-                Validator validator = schema.newValidator();
-                Source source = new StreamSource( new ByteArrayInputStream( xmlResult.getBytes() ) );
-                assertSchemaValid( validator, source );
-            }
+            boolean foundValidMetadataEntry = hasValidEntry( resultSet );
+            assertTrue( foundValidMetadataEntry,
+                        "GeoPackage does not contain at least one metadata entry valid against NIMS." );
         }
     }
 
-    private Schema createSchema() {
-        SchemaFactory schemaFactory = SchemaFactory.newInstance( XMLConstants.W3C_XML_SCHEMA_NS_URI );
-        try {
-            URL resource = getClass().getClassLoader().getResource( "org/opengis/cite/gpkg12/nsg/metadata/NMIS_v2.X_Schema/nas/nmis.xsd" );
-            return schemaFactory.newSchema( resource );
-        } catch ( Exception e ) {
-            LOG.log( Level.SEVERE, "Could not load schema 'mis.xsd'", e );
+    boolean hasValidEntry( ResultSet resultSet )
+                            throws SQLException {
+        while ( resultSet.next() ) {
+            String metadataEntry = resultSet.getString( "metadata" );
+            boolean isValid = isEntryValid( metadataEntry );
+            if ( isValid )
+                return true;
         }
-        return null;
+        return false;
+    }
+
+    private boolean isEntryValid( String metadataEntry ) {
+        if ( metadataEntry == null || metadataEntry.isEmpty() )
+            return false;
+        return isSchemaValid( metadataEntry );
+    }
+
+    private boolean isSchemaValid( String metadataEntry ) {
+        try {
+            Validator validator = schema.newValidator();
+            Source source = new StreamSource( new ByteArrayInputStream( metadataEntry.getBytes() ) );
+            ValidationErrorHandler errHandler = new ValidationErrorHandler();
+            validator.setErrorHandler( errHandler );
+            validator.validate( source );
+            return errHandler.errorsDetected();
+        } catch ( Exception var4 ) {
+            return false;
+        }
     }
 
 }

--- a/src/main/java/org/opengis/cite/gpkg12/nsg/metadata/MetadataTests.java
+++ b/src/main/java/org/opengis/cite/gpkg12/nsg/metadata/MetadataTests.java
@@ -1,14 +1,23 @@
 package org.opengis.cite.gpkg12.nsg.metadata;
 
-import static org.testng.Assert.assertTrue;
+import static java.util.logging.Level.WARNING;
 
 import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.StringReader;
 import java.net.URL;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
 
 import javax.xml.XMLConstants;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
 import javax.xml.transform.Source;
 import javax.xml.transform.stream.StreamSource;
 import javax.xml.validation.Schema;
@@ -20,11 +29,16 @@ import org.opengis.cite.validation.ValidationErrorHandler;
 import org.testng.SkipException;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
+import org.xml.sax.InputSource;
 import org.xml.sax.SAXException;
 
 public class MetadataTests extends CommonFixture {
 
+    private final static Logger LOG = Logger.getLogger( MetadataTests.class.getName() );
+
     private Schema schema;
+
+    private DocumentBuilder builder;
 
     @BeforeClass
     public void initSchema()
@@ -32,6 +46,14 @@ public class MetadataTests extends CommonFixture {
         SchemaFactory schemaFactory = SchemaFactory.newInstance( XMLConstants.W3C_XML_SCHEMA_NS_URI );
         URL resource = getClass().getClassLoader().getResource( "org/opengis/cite/gpkg12/nsg/metadata/NMIS_v2.X_Schema/nas/nmis.xsd" );
         this.schema = schemaFactory.newSchema( resource );
+    }
+
+    @BeforeClass
+    public void initDocumentBuilder()
+                            throws ParserConfigurationException {
+        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+        factory.setValidating( false );
+        this.builder = factory.newDocumentBuilder();
     }
 
     /**
@@ -48,40 +70,81 @@ public class MetadataTests extends CommonFixture {
 
         try (final Statement statement = this.databaseConnection.createStatement();
                                 final ResultSet resultSet = statement.executeQuery( "SELECT metadata FROM gpkg_metadata;" )) {
-            boolean foundValidMetadataEntry = hasValidEntry( resultSet );
-            assertTrue( foundValidMetadataEntry,
-                        "GeoPackage does not contain at least one metadata entry valid against NIMS." );
+            List<String> xmlEntries = findXmlEntries( resultSet );
+            validateXmlEntries( xmlEntries );
         }
     }
 
-    boolean hasValidEntry( ResultSet resultSet )
+    void validateXmlEntries( List<String> xmlEntries ) {
+        if ( xmlEntries.isEmpty() )
+            throw new AssertionError( "GeoPackage does not contain a NMIS v2.2 metadata as XML." );
+        if ( xmlEntries.size() == 1 ) {
+            validateXmlEntry( xmlEntries.get( 0 ) );
+        } else {
+            boolean hasValidMetadataEntry = hasValidEntry( xmlEntries );
+            if ( !hasValidMetadataEntry )
+                throw new AssertionError( "GeoPackage does not contain at least one valid NMIS v2.2 metadata as XML." );
+        }
+    }
+
+    private void validateXmlEntry( String xmlEntry ) {
+        try {
+            ValidationErrorHandler validationResult = validate( xmlEntry );
+            if ( validationResult.errorsDetected() ) {
+                String errors = validationResult.getErrors().stream().map( Object::toString ).collect( Collectors.joining( "," ) );
+                throw new AssertionError(
+                                          "GeoPackage does not contain a valid NMIS v2.2 metadata as XML. Results of the schema validation: "
+                                                                  + errors );
+            }
+        } catch ( IOException | SAXException e ) {
+            LOG.log( WARNING, "An error occurred during validation.", e );
+            throw new AssertionError( "Validation of the NMIS v2.2 metadata as XML failed: " + e.getMessage() );
+        }
+    }
+
+    List<String> findXmlEntries( ResultSet resultSet )
                             throws SQLException {
+        List<String> xmlEntries = new ArrayList<>();
         while ( resultSet.next() ) {
             String metadataEntry = resultSet.getString( "metadata" );
-            boolean isValid = isEntryValid( metadataEntry );
-            if ( isValid )
+            try (StringReader reader = new StringReader( metadataEntry )) {
+                InputSource inputSource = new InputSource( reader );
+                this.builder.parse( inputSource );
+                xmlEntries.add( metadataEntry );
+            } catch ( SAXException | IOException e ) {
+            }
+        }
+        return xmlEntries;
+    }
+
+    private boolean hasValidEntry( List<String> metadataEntries ) {
+        for ( String metadataEntry : metadataEntries ) {
+            if ( isEntryValid( metadataEntry ) )
                 return true;
         }
         return false;
     }
 
     private boolean isEntryValid( String metadataEntry ) {
-        if ( metadataEntry == null || metadataEntry.isEmpty() )
-            return false;
-        return isSchemaValid( metadataEntry );
+        try {
+            ValidationErrorHandler validationResult = validate( metadataEntry );
+            if ( !validationResult.errorsDetected() ) {
+                return true;
+            }
+        } catch ( IOException | SAXException e ) {
+            LOG.log( WARNING, "An error occurred during validation.", e );
+        }
+        return false;
     }
 
-    private boolean isSchemaValid( String metadataEntry ) {
-        try {
-            Validator validator = schema.newValidator();
-            Source source = new StreamSource( new ByteArrayInputStream( metadataEntry.getBytes() ) );
-            ValidationErrorHandler errHandler = new ValidationErrorHandler();
-            validator.setErrorHandler( errHandler );
-            validator.validate( source );
-            return errHandler.errorsDetected();
-        } catch ( Exception var4 ) {
-            return false;
-        }
+    private ValidationErrorHandler validate( String metadataEntry )
+                            throws IOException, SAXException {
+        Validator validator = schema.newValidator();
+        ValidationErrorHandler errHandler = new ValidationErrorHandler();
+        validator.setErrorHandler( errHandler );
+        Source source = new StreamSource( new ByteArrayInputStream( metadataEntry.getBytes() ) );
+        validator.validate( source );
+        return errHandler;
     }
 
 }

--- a/src/test/java/org/opengis/cite/gpkg12/nsg/metadata/MetadataTestsTest.java
+++ b/src/test/java/org/opengis/cite/gpkg12/nsg/metadata/MetadataTestsTest.java
@@ -11,48 +11,61 @@ import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
 import java.sql.ResultSet;
 import java.sql.SQLException;
+import java.util.List;
 
+import org.junit.Before;
 import org.junit.Test;
+import org.w3c.dom.Document;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.ParserConfigurationException;
 
 /**
  * @author <a href="mailto:goltz@lat-lon.de">Lyn Goltz </a>
  */
 public class MetadataTestsTest {
 
-    @Test
+    private MetadataTests metadataTests;
+
+    @Before
+    public void initMetadataTests()
+            throws SAXException, ParserConfigurationException {
+        this.metadataTests = new MetadataTests();
+        metadataTests.initSchema();
+        metadataTests.initDocumentBuilder();
+    }
+
+    @Test(expected = AssertionError.class)
     public void testWithOneInvalidEntry()
                             throws Exception {
-        MetadataTests metadataTests = new MetadataTests();
-        metadataTests.initSchema();
+        ResultSet resultSet = mockResultSetWithOneStringEntry();
+        List<String> xmlEntries = metadataTests.findXmlEntries( resultSet );
+        assertThat( xmlEntries.size(), is( 0 ) );
 
-        ResultSet resultSet = mockResultSetWithOneInvalidEntry();
-        boolean isValid = metadataTests.hasValidEntry( resultSet );
-        assertThat( isValid, is( false ) );
+        metadataTests.validateXmlEntries( xmlEntries );
     }
 
     @Test
     public void testWithOneValidEntry()
                             throws Exception {
-        MetadataTests metadataTests = new MetadataTests();
-        metadataTests.initSchema();
-
         ResultSet resultSet = mockResultSetWithOneValidEntry();
-        boolean isValid = metadataTests.hasValidEntry( resultSet );
-        assertThat( isValid, is( true ) );
+        List<String> xmlEntries = metadataTests.findXmlEntries( resultSet );
+        assertThat( xmlEntries.size(), is( 1 ) );
+
+        metadataTests.validateXmlEntries( xmlEntries );
     }
 
     @Test
     public void testWithOneInvalidAndOneValidEntry()
                             throws Exception {
-        MetadataTests metadataTests = new MetadataTests();
-        metadataTests.initSchema();
-
         ResultSet resultSet = mockResultSetWithOneInvalidAndOneValidEntry();
-        boolean isValid = metadataTests.hasValidEntry( resultSet );
-        assertThat( isValid, is( true ) );
+        List<String> xmlEntries = metadataTests.findXmlEntries( resultSet );
+        assertThat( xmlEntries.size(), is( 1 ) );
+
+        metadataTests.validateXmlEntries( xmlEntries );
     }
 
-    private ResultSet mockResultSetWithOneInvalidEntry()
+    private ResultSet mockResultSetWithOneStringEntry()
                             throws SQLException {
         ResultSet mock = mock( ResultSet.class );
         when( mock.next() ).thenReturn( true, false );

--- a/src/test/java/org/opengis/cite/gpkg12/nsg/metadata/MetadataTestsTest.java
+++ b/src/test/java/org/opengis/cite/gpkg12/nsg/metadata/MetadataTestsTest.java
@@ -1,0 +1,100 @@
+package org.opengis.cite.gpkg12.nsg.metadata;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+import org.junit.Test;
+
+/**
+ * @author <a href="mailto:goltz@lat-lon.de">Lyn Goltz </a>
+ */
+public class MetadataTestsTest {
+
+    @Test
+    public void testWithOneInvalidEntry()
+                            throws Exception {
+        MetadataTests metadataTests = new MetadataTests();
+        metadataTests.initSchema();
+
+        ResultSet resultSet = mockResultSetWithOneInvalidEntry();
+        boolean isValid = metadataTests.hasValidEntry( resultSet );
+        assertThat( isValid, is( false ) );
+    }
+
+    @Test
+    public void testWithOneValidEntry()
+                            throws Exception {
+        MetadataTests metadataTests = new MetadataTests();
+        metadataTests.initSchema();
+
+        ResultSet resultSet = mockResultSetWithOneValidEntry();
+        boolean isValid = metadataTests.hasValidEntry( resultSet );
+        assertThat( isValid, is( true ) );
+    }
+
+    @Test
+    public void testWithOneInvalidAndOneValidEntry()
+                            throws Exception {
+        MetadataTests metadataTests = new MetadataTests();
+        metadataTests.initSchema();
+
+        ResultSet resultSet = mockResultSetWithOneInvalidAndOneValidEntry();
+        boolean isValid = metadataTests.hasValidEntry( resultSet );
+        assertThat( isValid, is( true ) );
+    }
+
+    private ResultSet mockResultSetWithOneInvalidEntry()
+                            throws SQLException {
+        ResultSet mock = mock( ResultSet.class );
+        when( mock.next() ).thenReturn( true, false );
+        when( mock.getString( "metadata" ) ).thenReturn( simpleStringMetadata() );
+        return mock;
+    }
+
+    private ResultSet mockResultSetWithOneValidEntry()
+                            throws SQLException, IOException {
+        ResultSet mock = mock( ResultSet.class );
+        when( mock.next() ).thenReturn( true, false );
+        when( mock.getString( "metadata" ) ).thenReturn( validMetadata() );
+        return mock;
+    }
+
+    private ResultSet mockResultSetWithOneInvalidAndOneValidEntry()
+                            throws SQLException, IOException {
+        ResultSet mock = mock( ResultSet.class );
+        when( mock.next() ).thenReturn( true, true, false );
+        when( mock.getString( "metadata" ) ).thenReturn( emptyMetadata(), validMetadata() );
+        return mock;
+    }
+
+    private String validMetadata()
+                            throws IOException {
+        InputStream metadata = getClass().getResourceAsStream( "nmisExampleInstanceShort_ISM6.xml" );
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+        int nRead;
+        byte[] data = new byte[1024];
+        while ( ( nRead = metadata.read( data, 0, data.length ) ) != -1 ) {
+            buffer.write( data, 0, nRead );
+        }
+        buffer.flush();
+        return new String( buffer.toByteArray(), StandardCharsets.UTF_8 );
+    }
+
+    private String simpleStringMetadata() {
+        return "invalid";
+    }
+
+    private String emptyMetadata() {
+        return "";
+    }
+
+}

--- a/src/test/resources/org/opengis/cite/gpkg12/nsg/metadata/nmisExampleInstanceShort_ISM6.xml
+++ b/src/test/resources/org/opengis/cite/gpkg12/nsg/metadata/nmisExampleInstanceShort_ISM6.xml
@@ -13,7 +13,7 @@
 <!--<?xml-stylesheet href="./nas/nmisGmdSchem.xsl" type="text/xsl"?>-->
 <!--<?xml-stylesheet href="./nas/nmisGmiSchem.xsl" type="text/xsl"?>-->
 
-<nas:MD_Metadata xmlns:nas="http://metadata.ces.mil/dse/ns/GSIP/5.0/nas" xmlns:gmd="http://www.isotc211.org/2005/gmd" xmlns:gco="http://www.isotc211.org/2005/gco" xmlns:gml="http://www.opengis.net/gml/3.2" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ism="urn:us:gov:ic:ism" xmlns:ntk="urn:us:gov:ic:ntk" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://metadata.ces.mil/dse/ns/GSIP/5.0/nas ./nas/nmis.xsd" ism:DESVersion="6" ntk:DESVersion="4" ism:resourceElement="true" ism:createDate="2010-09-21" ism:classification="U" ism:ownerProducer="USA">
+<nas:MD_Metadata xmlns:nas="http://metadata.ces.mil/dse/ns/GSIP/5.0/nas" xmlns:gmd="http://www.isotc211.org/2005/gmd" xmlns:gco="http://www.isotc211.org/2005/gco" xmlns:gml="http://www.opengis.net/gml/3.2" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ism="urn:us:gov:ic:ism" xmlns:ntk="urn:us:gov:ic:ntk" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://metadata.ces.mil/dse/ns/GSIP/5.0/nas ./NMIS_v2.X_Schema/nas/nmis.xsd" ism:DESVersion="9" ntk:DESVersion="7" ism:resourceElement="true" ism:createDate="2010-09-21" ism:classification="U" ism:ownerProducer="USA">
     <!-- Note that the value of the xsi:schemaLocation in this example references a local copy of the nmis.xsd.
           The correct reference in the DSE is: http://metadata.ces.mil/dse/ns/GSIP/nmis/2.1.0/nmis.xsd.
   -->

--- a/src/test/resources/org/opengis/cite/gpkg12/nsg/metadata/nmisExampleInstanceShort_ISM6.xml
+++ b/src/test/resources/org/opengis/cite/gpkg12/nsg/metadata/nmisExampleInstanceShort_ISM6.xml
@@ -1,0 +1,366 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Example nmisExampleInstanceShort_ISM6.xml from https://nsgreg.nga.mil/NSGDOC/files/doc/Document/NMISv2.2SchemaApril2015.zip -->
+<!--<?oxygen SCHSchema="./nas/nmisGmd.sch"?>-->
+<!--<?oxygen SCHSchema="./nas/nmis.sch"?>-->
+<!--<?oxygen SCHSchema="./nas/nmisGmdProfileExclude.sch"?>-->
+<!--<?oxygen SCHSchema="./nas/nmisGmdProfileRestrict.sch"?>-->
+<!--<?oxygen SCHSchema="./nas/nmisGmiProfileExclude.sch"?>-->
+<!--<?oxygen SCHSchema="./nas/nmisGmiProfileRestrict.sch"?>-->
+<!--<?oxygen SCHSchema="./nas/nmisGmi.sch"?>-->
+<!--<?xml-stylesheet href="./nas/nmisSchem.xsl" type="text/xsl"?>-->
+<!--<?xml-stylesheet href="./nas/nmisGmdProfileRestrictSchem.xsl" type="text/xsl"?>-->
+<!--<?xml-stylesheet href="./nas/nmisGmdProfileExcludeSchem.xsl" type="text/xsl"?>-->
+<!--<?xml-stylesheet href="./nas/nmisGmdSchem.xsl" type="text/xsl"?>-->
+<!--<?xml-stylesheet href="./nas/nmisGmiSchem.xsl" type="text/xsl"?>-->
+
+<nas:MD_Metadata xmlns:nas="http://metadata.ces.mil/dse/ns/GSIP/5.0/nas" xmlns:gmd="http://www.isotc211.org/2005/gmd" xmlns:gco="http://www.isotc211.org/2005/gco" xmlns:gml="http://www.opengis.net/gml/3.2" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ism="urn:us:gov:ic:ism" xmlns:ntk="urn:us:gov:ic:ntk" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://metadata.ces.mil/dse/ns/GSIP/5.0/nas ./nas/nmis.xsd" ism:DESVersion="6" ntk:DESVersion="4" ism:resourceElement="true" ism:createDate="2010-09-21" ism:classification="U" ism:ownerProducer="USA">
+    <!-- Note that the value of the xsi:schemaLocation in this example references a local copy of the nmis.xsd.
+          The correct reference in the DSE is: http://metadata.ces.mil/dse/ns/GSIP/nmis/2.1.0/nmis.xsd.
+  -->
+    <!-- This minimum example describes a 2-dimensional vector dataset using the WGS84 ellipsoidal coordinate reference system encoded using ESRI Shapefiles. The dataset covers an area of southern New Jersey, USA and contains a sample of Local Topographic Data Store (LTDS) data for a specific cell that is part of a larger multi-cell dataset. The dataset was distributed by NGA on August 30, 2010. The data in the cell is current as of January 1, 2009 (temporal extent). Descriptive metadata is provided listing thematic keywords for the dataset. Associated geopolitical regions (i.e., countries and their subdivisions) are documented using "geographic descriptions". The dataset contains a limited number of feature types and those are defined in the NSG Entity Catalogue v5.0. Both the dataset and metadata are unclassified.
+  -->
+    <!-- Note that it is necessary to add XML attributes to the root node in order to correctly support DES.ISM.XML. The root node of any DES.ISM.XML-conformant instance document must specify the version of DES.ISM.XML used. Additionally, there must be a node that is identified as the "resource node" that would be used to generate banner marks and the E.O. 12958 classification authority block. The MD_Metadata node serves as the "resource node" and thus must carry, at a minimum, the necessary identification, createDate, classification, and ownerProducer XML attributes.
+
+  - The value of createDate shall be the same as that for the <gmd:dateStamp><gco:Date> XML element.
+  - The values of classification, ownerProducer and any additional DES.ISM.XML attributes shall be based on those assigned for the corresponding gmd:metadataConstraints/nas:MD_SecurityConstraints/nas:capcoMarking and gmd:identificationInfo/nas:MD_DataIdentification/gmd:resourceConstraints/nas:MD_SecurityConstraints/nas:capcoMarking XML attributes as determined by applicable banner and E.O. 12958 classification authority block marking requirements.
+  -->
+    <!-- Note that it is necessary to add XML attributes to the root node in order to correctly support DES.NTK.XML. The root node of any DES.NTK.XML-conformant instance document must specify the version of DES.NTK.XML used.
+  -->
+    <!-- Data Content Hierarchy Level: In this example we suppose that this metadata record documents one cell
+          of LTDS data that is a member of a larger dataset consisting of 16 cells. -->
+    <gmd:hierarchyLevel>
+        <gmd:MD_ScopeCode codeList="http://api.nsgreg.nga.mil/codelist/ScopeCode" codeListValue="dataset"/>
+    </gmd:hierarchyLevel>
+    <gmd:hierarchyLevelName>
+        <nas:ScopeAmplificationCode codeList="http://api.nsgreg.nga.mil/codelist/ScopeAmplificationCode" codeListValue="cell"/>
+    </gmd:hierarchyLevelName>
+    <!-- Resource Metadata Contact: This is the preparer of this metadata record. Since the same organization that created the
+          data sample also created this metadata record we simply reference their specification elsewhere in this metadata record. -->
+    <gmd:contact>
+        <gmd:CI_ResponsibleParty>
+            <gmd:organisationName>
+                <gco:CharacterString>US National Geospatial-Intelligence Agency</gco:CharacterString>
+            </gmd:organisationName>
+            <gmd:role>
+                <gmd:CI_RoleCode codeList="http://api.nsgreg.nga.mil/codelist/RoleCode" codeListValue="originator"/>
+            </gmd:role>
+        </gmd:CI_ResponsibleParty>
+    </gmd:contact>
+    <!-- Resource Metadata Date: This is the date that this metadata record was prepared (and its content known to be accurate).  -->
+    <gmd:dateStamp>
+        <gco:Date>2010-09-21</gco:Date>
+    </gmd:dateStamp>
+    <!-- Resource Metadata Standard: This is the metadata specification with regards to which this metadata record is valid.
+          It is a Recommended Best Practice to use the full name of the metadata standard in order to eliminate ambiguity.
+          It may be the case that an extension of the NMIS was used and therefore these values may vary.  -->
+    <gmd:metadataStandardName>
+        <nas:MetadataStandardNameCode codeList="http://api.nsgreg.nga.mil/codelist/MetadataStandardNameCode" codeListValue="nsgMetadataFoundation"/>
+    </gmd:metadataStandardName>
+    <gmd:metadataStandardVersion>
+        <nas:MetadataStandardVersion>2.2.0</nas:MetadataStandardVersion>
+    </gmd:metadataStandardVersion>
+    <!-- Resource Metadata String Localization Information: It is a Recommended Best Practice to explicitly specify
+    the "locale" of all string-values used in this metadata record. If this information is omitted then it shall be the
+    case that the string-values are all specified in terms of U.S. English using the UTF-8 character encoding.  -->
+    <gmd:locale>
+        <gmd:PT_Locale>
+            <gmd:languageCode>
+                <gmd:LanguageCode codeList="http://api.nsgreg.nga.mil/codelist/ISO639-2" codeListValue="eng"/>
+            </gmd:languageCode>
+            <gmd:country>
+                <gmd:Country codeList="http://api.nsgreg.nga.mil/geo-political/GENC/3/2-1" codeListValue="USA"/>
+            </gmd:country>
+            <gmd:characterEncoding>
+                <gmd:MD_CharacterSetCode codeList="http://api.nsgreg.nga.mil/codelist/CharacterSetCode" codeListValue="utf8"/>
+            </gmd:characterEncoding>
+        </gmd:PT_Locale>
+    </gmd:locale>
+    <!-- Data Content Reference System(s) Information: This specifies the spatial-temporal reference system(s) used
+          by the Data Content described by this metadata record. These shall always be specified based on the
+          dynamic resources hosted in the DoD Metadata Registry at: http://metadata.ces.mil/mdr/ns/GSIP/crs  -->
+    <gmd:referenceSystemInfo>
+        <gmd:MD_ReferenceSystem>
+            <gmd:referenceSystemIdentifier>
+                <gmd:RS_Identifier>
+                    <gmd:code>
+                        <gco:CharacterString>WGS84E_2D</gco:CharacterString>
+                    </gmd:code>
+                    <gmd:codeSpace>
+                        <gco:CharacterString>http://api.nsgreg.nga.mil/coord-ref-system</gco:CharacterString>
+                    </gmd:codeSpace>
+                </gmd:RS_Identifier>
+            </gmd:referenceSystemIdentifier>
+        </gmd:MD_ReferenceSystem>
+    </gmd:referenceSystemInfo>
+    <!-- Data Content Identification Information: The metadata specifically about the Data Content.  -->
+    <gmd:identificationInfo>
+        <nas:MD_DataIdentification>
+            <!-- Both an informal title (with distribution date) as well as a formal identifier is specified for the data resource content.  -->
+            <gmd:citation>
+                <gmd:CI_Citation>
+                    <gmd:title>
+                        <gco:CharacterString>Local Topographic Data Store (LTDS) data sample for cell 39N078WA.</gco:CharacterString>
+                    </gmd:title>
+                    <gmd:date>
+                        <gmd:CI_Date>
+                            <gmd:date>
+                                <gco:Date>2010-08-30</gco:Date>
+                            </gmd:date>
+                            <gmd:dateType>
+                                <gmd:CI_DateTypeCode codeList="http://api.nsgreg.nga.mil/codelist/DateTypeCode" codeListValue="distribution"/>
+                            </gmd:dateType>
+                        </gmd:CI_Date>
+                    </gmd:date>
+                </gmd:CI_Citation>
+            </gmd:citation>
+            <gmd:abstract>
+                <gco:CharacterString>A small-footprint sample of NAS-conformant data located in the southern region of New Jersey, USA. It contains a variety of urban, terrain, shoreline, and littoral vector-based feature data, some of which are vertical obstructions to aircraft.</gco:CharacterString>
+            </gmd:abstract>
+            <gmd:pointOfContact>
+                <gmd:CI_ResponsibleParty>
+                    <gmd:organisationName>
+                        <gco:CharacterString>US National Geospatial-Intelligence Agency</gco:CharacterString>
+                    </gmd:organisationName>
+                    <gmd:contactInfo>
+                        <gmd:CI_Contact>
+                            <gmd:phone>
+                                <gmd:CI_Telephone>
+                                    <gmd:voice>
+                                        <!-- An “internationalized” telephone number as specified in ITU-T Recommendation E.123 Clause 2. -->
+                                        <nas:TelephoneNumber>+1 703 555 1212</nas:TelephoneNumber>
+                                    </gmd:voice>
+                                </gmd:CI_Telephone>
+                            </gmd:phone>
+                            <gmd:address>
+                                <gmd:CI_Address>
+                                    <gmd:deliveryPoint>
+                                        <gco:CharacterString>Suite 100</gco:CharacterString>
+                                    </gmd:deliveryPoint>
+                                    <gmd:deliveryPoint>
+                                        <gco:CharacterString>123 Main St.</gco:CharacterString>
+                                    </gmd:deliveryPoint>
+                                    <gmd:city>
+                                        <gco:CharacterString>Chantilly</gco:CharacterString>
+                                    </gmd:city>
+                                    <gmd:administrativeArea>
+                                        <gco:CharacterString>VA</gco:CharacterString>
+                                    </gmd:administrativeArea>
+                                    <gmd:postalCode>
+                                        <gco:CharacterString>20151</gco:CharacterString>
+                                    </gmd:postalCode>
+                                    <gmd:country>
+                                        <gco:CharacterString>US</gco:CharacterString>
+                                    </gmd:country>
+                                </gmd:CI_Address>
+                            </gmd:address>
+                        </gmd:CI_Contact>
+                    </gmd:contactInfo>
+                    <gmd:role>
+                        <gmd:CI_RoleCode codeList="http://api.nsgreg.nga.mil/codelist/RoleCode" codeListValue="originator"/>
+                    </gmd:role>
+                </gmd:CI_ResponsibleParty>
+            </gmd:pointOfContact>
+            <!-- Data Content Thematic Keyword(s): Two thematic keywords are identified from a specified thesaurus.  -->
+            <gmd:descriptiveKeywords>
+                <gmd:MD_Keywords>
+                    <gmd:keyword>
+                        <gco:CharacterString>Boundaries and Lines</gco:CharacterString>
+                    </gmd:keyword>
+                    <gmd:keyword>
+                        <gco:CharacterString>General Structures</gco:CharacterString>
+                    </gmd:keyword>
+                    <gmd:type>
+                        <gmd:MD_KeywordTypeCode codeList="http://api.nsgreg.nga.mil/codelist/KeywordTypeCode" codeListValue="theme"/>
+                    </gmd:type>
+                    <gmd:thesaurusName>
+                        <gmd:CI_Citation>
+                            <gmd:title>
+                                <gco:CharacterString>NSG Entity Catalog View Thesaurus v5.0</gco:CharacterString>
+                            </gmd:title>
+                            <gmd:date>
+                                <gmd:CI_Date>
+                                    <gmd:date>
+                                        <gco:Date>2012-10-01</gco:Date>
+                                    </gmd:date>
+                                    <gmd:dateType>
+                                        <gmd:CI_DateTypeCode codeList="http://api.nsgreg.nga.mil/codelist/DateTypeCode" codeListValue="publication"/>
+                                    </gmd:dateType>
+                                </gmd:CI_Date>
+                            </gmd:date>
+                        </gmd:CI_Citation>
+                    </gmd:thesaurusName>
+                </gmd:MD_Keywords>
+            </gmd:descriptiveKeywords>
+            <!-- Data Content Security Constraint(s): This specifies any restrictions applicable to the data content itself.
+              It is required that the CAPCO Marking information be specified and that the corresponding value of the ISO 19115
+              classification and classificationSystem be specified so that in the case that the CAPCO Marking is removed
+              that the remaining security constraint specification be unambiguous.  -->
+            <gmd:resourceConstraints>
+                <nas:MD_SecurityConstraints>
+                    <gmd:classification>
+                        <gmd:MD_ClassificationCode codeList="http://api.nsgreg.nga.mil/codelist/ClassificationCode" codeListValue="unclassified"/>
+                    </gmd:classification>
+                    <!-- In the case of CAPCO markings the classification system must be "US CAPCO"
+                and the DES.ISM.XML attributes populated accordingly.  -->
+                    <gmd:classificationSystem>
+                        <nas:ClassificationSystem>US CAPCO</nas:ClassificationSystem>
+                    </gmd:classificationSystem>
+                    <nas:capcoMarking ism:classification="U" ism:ownerProducer="USA"/>
+                </nas:MD_SecurityConstraints>
+            </gmd:resourceConstraints>
+            <!-- Data Content String Localization Information: It is a Recommended Best Practice to explicitly specify
+              the "locale" of all string-values used in the Data Content resource. In the NSG it is anticipated that all data
+              resources will at least employ U.S. English using the UTF-8 character encoding for specifying string-values,
+              however in a coalition environment or where data content is based on open sources and/or legacy data
+              resources, other languages, country-specific language localizations, and character sets may be employed.  -->
+            <gmd:language>
+                <gmd:LanguageCode codeList="http://api.nsgreg.nga.mil/codelist/ISO639-2" codeListValue="eng"/>
+            </gmd:language>
+            <gmd:characterSet>
+                <gmd:MD_CharacterSetCode codeList="http://api.nsgreg.nga.mil/codelist/CharacterSetCode" codeListValue="utf8"/>
+            </gmd:characterSet>
+            <!-- Data Content Topic Category(ies): As specified by ISO 19115, these topic categories furnish a very high-level
+              means to discover geospatial data across the entire international community, including coalition forces/systems.  -->
+            <gmd:topicCategory>
+                <gmd:MD_TopicCategoryCode>boundaries</gmd:MD_TopicCategoryCode>
+            </gmd:topicCategory>
+            <gmd:topicCategory>
+                <gmd:MD_TopicCategoryCode>intelligenceMilitary</gmd:MD_TopicCategoryCode>
+            </gmd:topicCategory>
+            <gmd:topicCategory>
+                <gmd:MD_TopicCategoryCode>inlandWaters</gmd:MD_TopicCategoryCode>
+            </gmd:topicCategory>
+            <gmd:topicCategory>
+                <gmd:MD_TopicCategoryCode>location</gmd:MD_TopicCategoryCode>
+            </gmd:topicCategory>
+            <gmd:topicCategory>
+                <gmd:MD_TopicCategoryCode>oceans</gmd:MD_TopicCategoryCode>
+            </gmd:topicCategory>
+            <gmd:topicCategory>
+                <gmd:MD_TopicCategoryCode>structure</gmd:MD_TopicCategoryCode>
+            </gmd:topicCategory>
+            <gmd:topicCategory>
+                <gmd:MD_TopicCategoryCode>transportation</gmd:MD_TopicCategoryCode>
+            </gmd:topicCategory>
+            <gmd:topicCategory>
+                <gmd:MD_TopicCategoryCode>utilitiesCommunication</gmd:MD_TopicCategoryCode>
+            </gmd:topicCategory>
+            <!--Data Content Spatio-temporal Extent: Specify the extent of the Data Content in terms of a minimum
+            bounding rectangle (MBR) within which the data content is spatially located, however for greater
+            accuracy a polygon specifying the exact extent within that MBR is also specified.  -->
+            <gmd:extent>
+                <gmd:EX_Extent>
+                    <!-- Note that, as specified by ISO 19115, this is only an approximate reference intended principally
+                  for the purposes of data discovery therefore specifying the coordinate reference system is unnecessary.
+                  It is a Recommended Best Practice, however, that the WGS84 Ellipsoid be used as the
+                  horizontal datum when preparing values for this use.  -->
+                    <gmd:geographicElement>
+                        <gmd:EX_GeographicBoundingBox>
+                            <gmd:westBoundLongitude>
+                                <gco:Decimal>-74.92269</gco:Decimal>
+                            </gmd:westBoundLongitude>
+                            <gmd:eastBoundLongitude>
+                                <gco:Decimal>-74.66142</gco:Decimal>
+                            </gmd:eastBoundLongitude>
+                            <gmd:southBoundLatitude>
+                                <gco:Decimal>39.10306</gco:Decimal>
+                            </gmd:southBoundLatitude>
+                            <gmd:northBoundLatitude>
+                                <gco:Decimal>39.19414</gco:Decimal>
+                            </gmd:northBoundLatitude>
+                        </gmd:EX_GeographicBoundingBox>
+                    </gmd:geographicElement>
+                    <!-- Other informal methods may be used to describe the spatial-temporal extent of the data resource content.
+                  One approach leverages ISO 3166 as a gazetteer of countries and their administrative divisions.  -->
+                    <gmd:geographicElement>
+                        <gmd:EX_GeographicDescription>
+                            <gmd:geographicIdentifier>
+                                <!-- Reference to the ISO 3166-2 standard for codes of administrative divisions. The referenced authority
+                          is a NSG Registry-hosted online resource against which the specified code can be dynamically verified
+                          (and its meaning determined).  -->
+                                <gmd:MD_Identifier>
+                                    <gmd:authority xlink:href="http://api.nsgreg.nga.mil/geo-division/GENC/6/2-1"/>
+                                    <gmd:code>
+                                        <gco:CharacterString>US-NJ</gco:CharacterString>
+                                    </gmd:code>
+                                </gmd:MD_Identifier>
+                            </gmd:geographicIdentifier>
+                        </gmd:EX_GeographicDescription>
+                    </gmd:geographicElement>
+                    <!-- Specify that the data resource content was valid for the time period ending 1 January 2010.  -->
+                    <gmd:temporalElement>
+                        <gmd:EX_TemporalExtent>
+                            <gmd:extent>
+                                <gml:TimePeriod gml:id="ResourceTemporalExtent">
+                                    <gml:begin>
+                                        <gml:TimeInstant gml:id="ResourceTemporalExtentBegin">
+                                            <gml:timePosition indeterminatePosition="unknown"/>
+                                        </gml:TimeInstant>
+                                    </gml:begin>
+                                    <gml:end>
+                                        <gml:TimeInstant gml:id="ResourceTemporalExtentEnd">
+                                            <gml:timePosition>2010-01-02</gml:timePosition>
+                                        </gml:TimeInstant>
+                                    </gml:end>
+                                </gml:TimePeriod>
+                            </gmd:extent>
+                        </gmd:EX_TemporalExtent>
+                    </gmd:temporalElement>
+                    <!-- Specify the vertical extent of the data resource content as starting at -100.0 and extending through 1500.0
+                  metres (below/above) the WGS84 Earth Gravity Model 1996 geoidal surface. This surface is considered
+                  to be equivalent to Mean Sea Level for many purposes. Note that the units of measure for the pair of
+                  real values is determined by the gmd:verticalCRS, which is specified in the NSG Registry online resource.  -->
+                    <gmd:verticalElement>
+                        <gmd:EX_VerticalExtent>
+                            <gmd:minimumValue>
+                                <gco:Real>-100.0</gco:Real>
+                            </gmd:minimumValue>
+                            <gmd:maximumValue>
+                                <gco:Real>1500.0</gco:Real>
+                            </gmd:maximumValue>
+                            <gmd:verticalCRS xlink:href="http://api.nsgreg.nga.mil/coord-ref-system/EGM96_H"/>
+                        </gmd:EX_VerticalExtent>
+                    </gmd:verticalElement>
+                </gmd:EX_Extent>
+            </gmd:extent>
+            <!-- Remaining Data Content String Localization Information: The last (third) component of the string-localization
+              triple of characteristics. Specifies the country of the language in order to be able to discriminate, e.g.,
+              U.S. English from U.K. English, or Canadian French from Senegalese French.  -->
+            <nas:languageCountry>
+                <nas:LanguageCountryCode codeList="http://api.nsgreg.nga.mil/geo-political/GENC/3/2-1" codeListValue="USA"/>
+            </nas:languageCountry>
+            <!-- new in NMIS v2.1; mandatory -->
+            <nas:resourceCategory>
+                <nas:ResourceCategoryCode codeList="http://api.nsgreg.nga.mil/codelist/ResourceCategoryCode" codeListValue="topographicFeatures"/>
+            </nas:resourceCategory>
+        </nas:MD_DataIdentification>
+    </gmd:identificationInfo>
+    <!-- Resource Metadata Constraint(s): This specifies any restrictions applicable to the content of this metadata record.
+          It is required that the CAPCO Marking information be specified and that the corresponding value of the ISO 19115
+          classification and classificationSystem be specified so that in the case that the CAPCO Marking is removed
+          that the remaining security constraint specification be unambiguous.  -->
+    <gmd:metadataConstraints>
+        <nas:MD_SecurityConstraints>
+            <gmd:classification>
+                <gmd:MD_ClassificationCode codeList="http://api.nsgreg.nga.mil/codelist/ClassificationCode" codeListValue="unclassified"/>
+            </gmd:classification>
+            <!-- In the case of CAPCO markings the classification system must be "US CAPCO"
+            and the DES.ISM.XML attributes populated accordingly.  -->
+            <gmd:classificationSystem>
+                <nas:ClassificationSystem>US CAPCO</nas:ClassificationSystem>
+            </gmd:classificationSystem>
+            <nas:capcoMarking ism:classification="U" ism:ownerProducer="USA"/>
+            <!-- Restrict use to the systems participating in the NSG who are members of the User group -->
+            <nas:needToKnow>
+                <ntk:Access ism:classification="U" ism:ownerProducer="USA">
+                    <ntk:AccessGroupList>
+                        <ntk:AccessGroup ism:classification="U" ism:ownerProducer="USA">
+                            <ntk:AccessSystemName ism:classification="U" ism:ownerProducer="USA">NSG</ntk:AccessSystemName>
+                            <ntk:AccessGroupValue ism:classification="U" ism:ownerProducer="USA">User</ntk:AccessGroupValue>
+                        </ntk:AccessGroup>
+                    </ntk:AccessGroupList>
+                </ntk:Access>
+            </nas:needToKnow>
+        </nas:MD_SecurityConstraints>
+    </gmd:metadataConstraints>
+</nas:MD_Metadata>

